pkg/controller/nodes/attr_path_resolver.go

```diff
@@ -0,0 +1,149 @@
+	}
+
+	// resolve dataclass
+	if currVal.GetScalar() != nil && currVal.GetScalar().GetGeneric() != nil {
+		st := currVal.GetScalar().GetGeneric()
+		// start from index "count"
+		currVal, err = resolveAttrPathInPbStruct(ctx, nodeID, st, bindAttrPath[count:])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return currVal, nil
+}
+
+func resolveAttrPathInPbStruct(ctx context.Context, nodeID string, st *structpb.Struct, bindAttrPath []*core.PromiseAtrribute) (*core.Literal, error) {
+
+	var currVal interface{}
+	var tmpVal interface{}
+	var exist bool
+
+	currVal = st.AsMap()
+
+	// Turn the current value to a map so it can be resolved more easily
+	for _, attr := range bindAttrPath {
+		switch currVal.(type) {
+		// map
+		case map[string]interface{}:
+			tmpVal, exist = currVal.(map[string]interface{})[attr.GetStringValue()]
+			if exist == false {
+				return nil, errors.Errorf(errors.PromiseAttributeResolveError, nodeID, "key [%v] does not exist in literal %v", attr.GetStringValue(), currVal)
+			}
+			currVal = tmpVal
+		// list
+		case []interface{}:
+			if int(attr.GetIntValue()) >= len(currVal.([]interface{})) {
+				return nil, errors.Errorf(errors.PromiseAttributeResolveError, nodeID, "index [%v] is out of range of %v", attr.GetIntValue(), currVal)
+			}
+			currVal = currVal.([]interface{})[attr.GetIntValue()]
+		}
+	}
+
+	literal := &core.Literal{}
+
+	// Different types of current value will be wrapped in different ways
+	switch currVal.(type) {
+	// We don't need to wrap map[string]interface{} to LiteralMap, because DictTransformer in flytekit can also handle pb struct
+	case map[string]interface{}:
+		new_st, err := structpb.NewStruct(currVal.(map[string]interface{}))
+		if err != nil {
+			return nil, err
+		}
+		literal.Value = &core.Literal_Scalar{
+			Scalar: &core.Scalar{
+				Value: &core.Scalar_Generic{
+					Generic: new_st,
+				},
+			},
+		}
+	case []interface{}:
+		literals := []*core.Literal{}
+		for _, v := range currVal.([]interface{}) {
+			literalItem := &core.Literal{}
+			scalar, err := resolveInterfaceToLiteralScalar(ctx, nodeID, v)
+			if err != nil {
+				return nil, err
+			}
+			literalItem.Value = scalar
+			literals = append(literals, literalItem)
+		}
+
+		literal.Value = &core.Literal_Collection{
+			Collection: &core.LiteralCollection{
+				Literals: literals,
+			},
+		}
+	case interface{}:
+		scalar, err := resolveInterfaceToLiteralScalar(ctx, nodeID, currVal)
+		if err != nil {
+			return nil, err
+		}
+		literal.Value = scalar
+	}
+
+	return literal, nil
+}
+
+func resolveInterfaceToLiteralScalar(ctx context.Context, nodeID string, obj interface{}) (*core.Literal_Scalar, error) {
+	value := &core.Primitive{}
+
+	switch obj.(type) {
+	case string:
+		value.Value = &core.Primitive_StringValue{StringValue: obj.(string)}
+	case int:
+		value.Value = &core.Primitive_Integer{Integer: int64(obj.(int))}
+	case float64:
+		value.Value = &core.Primitive_FloatValue{FloatValue: obj.(float64)}
+	case bool:
```

Comment:
How about the other types in the scalar? https://github.com/flyteorg/flyteidl/blob/7f62a157921db31d5f7a2285b4e9a42b4a01eaa4/protos/flyteidl/core/literals.proto#L84-L90